codeworld-server/src/ErrorSanitizer.hs

```diff
@@ -24,62 +24,69 @@ import           Data.ByteString (ByteString)
+import           Data.Tuple
 import           System.Directory
 import           System.FilePath
 import           System.IO
 import           System.IO.Temp (withSystemTempDirectory)
 import           System.Process
-import           Text.Regex.TDFA
+import           Text.Regex.TDFA hiding (Regex)
 import           Text.Regex
 
 import Util
 
+filterStages :: [(Regex, String)]
+filterStages = [
+        (mkRegex "\226\8364\162", ""),
+        (mkRegex "\226\8364\732", ""),
+        (mkRegex "\226\8364\8482", ""),
+        (mkRegex "\\’", ""),
+        (mkRegex "IO action main", "variable main"),
+        (mkRegex "module Main", "your program"),
+        (mkRegex "main:Main", "your program"),
+        (mkRegex "Couldn't match expected type Text\\s*with actual type GHC.Types.Char",
+        	"Text requires double quotes, rather than single."),
+        (mkRegex "base-[0-9.]*:GHC\\.Stack\\.Types\\.HasCallStack => ", ""),
+        (mkRegex "GHC\\.Types\\.Char" ,""),
+        (mkRegex "codeworld-base[-.:_A-Za-z0-9]*", "the standard library"),
+        (mkRegex "Main\\.", ""),
+        (mkRegex "main :: t", "main program"),
+        (mkRegex "Prelude\\.", ""),
+        (mkRegex "IO \\(\\)", "Program"),
+        (mkRegex "IO [a-z][a-zA-Z0-9_]*", "Program"),
+        (mkRegex "\\bBool\\b", "Truth"),
+        (mkRegex "[ ]*Perhaps you intended to use TemplateHaskell\n", ""),
+        (mkRegex "imported from [^)\n]*", "defined in the standard library"),
+        (mkRegex "the first argument", "the parameter(s)"),
+        (mkRegex "[ ]*The function [a-zA-Z0-9_]* is applied to [a-z0-9]* arguments,\n", ""),
+        (mkRegex "[ ]*but its type .* has only .*\n", ""),
+        (mkRegex "A data constructor of that name is in scope; did you mean DataKinds\\?",
+        	"That name refers to a value, not a type."),
+        (mkRegex "type constructor or class", "type"),
+        (mkRegex "Illegal tuple section: use TupleSections",
+        	"This tuple is missing a value, or has an extra comma."),
+        (mkRegex "Use -v to see a list of the files searched for\\.", ""),
+        (mkRegex "\n\\s+\n", "\n"),
+        (mkRegex "When checking that:\\s*[^\n]*\\s*is more polymorphic than:\\s*[^\n]*(\n\\s*)?",
+        	""),
+        (mkRegex "\\[GHC\\.Types\\.Char\\] -> ", "\n"),
+        (mkRegex "\\(and originally defined in [^)]*\\)", "\n"),
+        (mkRegex "in string\\/character literal", "in text literal"),
+        (mkRegex "CallStack \\(from HasCallStack\\):", "When Evaluating:"),
+        (mkRegex "at src\\/[A-Za-z0-9\\/.:]*", ""),
+        (mkRegex "GHC\\.[A-Za-z.]*(\\s|\n)*->( |\n)*", ""),
+        (mkRegex "integer-gmp(-[0-9\\.]*)?:(.|\n)*?->( |\n)*", ""),
+        (mkRegex "base(-[0-9.]*)?\\:(.|\n)*?->( |\n)*", "\n"),
+        (mkRegex "lexical error at character '\822[01]'", "Smart quotes are not allowed.")
+    ]
+
+foldrec :: [Char] -> [(Regex, String)] -> String
+foldrec z [] = z
+foldrec z (x:xs) = subRegex (fst x) (foldrec z xs) (snd x)
```

Comment:
You can write it yourself, if you prefer.  But it's important to use a left fold and be strict in the accumulator.  Otherwise, you'll unnecessarily build up data in memory.